README.md
# NeosVR-Headless-API

**NeosVR-Headless-API** is a Python wrapper and API for the NeosVR headless client. With it, you can programmatically control headless clients both locally and remotely.

It aims to be easy to use by doing the heavy work of parsing command output for you and turns it into Python objects that you can use any way you like.

The [wiki](https://github.com/glitchfur/NeosVR-Headless-API/wiki) has more information about NeosVR-Headless-API, or if you want to get started right away you can jump into [Getting started](https://github.com/glitchfur/NeosVR-Headless-API/wiki/Getting-started). The [online documentation](https://docs.glitchfur.net/NeosVR-Headless-API) has a detailed explanation of all the methods available and how to use them.

```python
In [1]: from neosvr_headless_api import LocalHeadlessClient

In [2]: hc = LocalHeadlessClient("/home/glitch/Servers/NeosVR")

In [3]: hc.wait_for_ready()
Out[3]: True

In [4]: hc.status()
Out[4]: 
{'name': "Glitch's World",
 'session_id': 'S-1462cabf-9ac4-46b7-80b4-3921455093e1',
 'current_users': 1,
 'present_users': 0,
 'max_users': 8,
 'uptime': '00:00:18.9775300',
 'access_level': 'Anyone',
 'hidden_from_listing': False,
 'mobile_friendly': False,
 'description': 'This is where a cat knocks a bunch of stuff over.',
 'tags': ["computer", "cat", "code", "coffee", "chaos"],
 'users': ['GlitchHost']}

In [5]: hc.shutdown()
Out[5]: 0
```

This project uses the [GNU GPLv3](/LICENSE.txt) license.


CHANGELOG.md
# Changelog

## 2022-06-23

* The project now uses the [GNU GPLv3](/LICENSE.txt) license.

## 2022-06-22

* Add `rpyc` and `pdoc` packages to requirements.txt, for the RPC server and generating technical documentation respectively.
* Overhaul README with an example of usage.
* Greatly improve source documentation with more detailed docstrings and type annotations.
* All other headless client commands not yet implemented now raise `NotImplementedError` when attempting to call them.

## 2021-07-10
* Relaxed timeouts for commands and RPC connections:
  * Commands will now time out after 30 seconds. (Changed from 20 seconds.)
  * RPC connections now have a 60 second timeout. (Changed from 30 seconds.)

## 2021-07-09
* Add `CommandTimeout` exception which is raised when a command takes too long to return any output.

## 2021-07-02
* Overhauled the shutdown methods of the headless client:
  * `shutdown()` now accepts an optional `timeout` keyword argument to specify how long to wait (in seconds) for the headless client to shut down before raising a `TimeoutExpired` exception. `wait` can also be set to `False` to make the call non-blocking, in which case `timeout` is ignored and returns immediately, but without an exit code.
  * `sigint()` can be called to issue a `SIGINT` signal to the headless client, which should emulate pressing `Ctrl+C` in the console and cleanly shut down the headless client. This also accepts the `timeout` and `wait` keyword arguments mentioned above. Returns `-2`, if `wait` is `True`.
  * `terminate()` can be called to issue a `SIGTERM` signal to the headless client. This also accepts the `timeout` and `wait` keyword arguments mentioned above. Returns `-15`, if `wait` is `True`.
  * `kill()` can be called to issue a `SIGKILL` signal to the headless client. This also accepts the `timeout` and `wait` keyword arguments mentioned above. Returns `-9`, if `wait` is `True`.
  * `wait_for_shutdown()` is a blocking call that will wait for the headless client to shut down. The `timeout` keyword argument can be specified to wait up to `timeout` seconds for the headless client to shut down before raising a `TimeoutExpired` exception. Returns the exit code. This can be used after using any of the above functions in non-blocking mode (when `wait` is `False`).
* New methods exist for waiting on the headless client's "ready" state:
  * **Breaking change:** `wait()` has been renamed to `wait_for_ready()` and now accepts a `timeout` keyword argument. Blocks and returns `True` when the headless client is ready to accept commands. If `timeout` is specified, the call will return `False` if the headless client does not enter a ready state within `timeout` seconds.
  * `is_ready()` can be called to immediately check whether the headless client is ready or not without blocking. Returns either `True` if it is ready, or `False` if it is not ready.
* The `users()` function will now show whether the users in a session are silenced or not, which is a feature that was added in Neos version [2021.7.1.437](https://github.com/Neos-Metaverse/NeosPublic/issues/2517#issuecomment-871997486).

## 2021-06-27
* Defined constants for Neos' roles and access levels.
  * For user roles, they are: `ADMIN`, `BUILDER`, `MODERATOR`, `GUEST`, and `SPECTATOR`.
  * For session/world access levels, they are: `PRIVATE`, `LAN`, `FRIENDS`, `FRIENDS_OF_FRIENDS`, `REGISTERED_USERS`, and `ANYONE`.

## 2021-06-23
* Added an argument parser to the RPC server. It now accepts the following arguments:
  * `--host`: Specify the host or IP to bind to. Defaults to `127.0.0.1` (previously `0.0.0.0`).
  * `-p` or `--port`: Specify the TCP port to bind to. Defaults to `16881`.

## 2021-06-21
* Fixed a conditional parsing bug in the `users()` function where a headless user would not show up in the user list if it was not logged into a Neos account. This occurred because unauthenticated headless users have no user IDs, and this would cause parsing to fail. Headless users will now have their user ID set to `None` in these instances.

## 2021-06-18
* User IDs are now included when listing the users in a session, which is a feature that was added in Neos version [2021.6.18.4](https://github.com/Neos-Metaverse/NeosPublic/issues/2485#issuecomment-863639775).

## 2021-06-17
* Added checks for new error messages added in Neos version [2021.6.14.1031](https://github.com/Neos-Metaverse/NeosPublic/issues/2457#issuecomment-860859990), which occur when attempting to use `banByName` or `banByID` to ban a user who has already been banned.

## 2021-06-16
* Added exception for unexpected errors that occur in the headless client.

## 2021-06-12
* Added checks for additional error messages, particularly "Not logged in!" errors that occur when running a command that requires a logged in user while not being logged in.

## 2021-06-11
* Parsing of commands is now much more strict and less error prone. See [this GitHub issue](https://github.com/Neos-Metaverse/NeosPublic/issues/2436) for information on why this is required.
* The `status` command in particular has much improved parsing:
  * The check for an empty session ID has been removed. This used to be needed as the Userspace world had no session ID, but the [Userspace world has been removed from the world listing](https://github.com/Neos-Metaverse/NeosPublic/issues/1811#issuecomment-793282104) as of Neos version 2021.3.9.130.
  * Worlds with blank names are now parsed properly. The name will be set to `None` if there is no name.
  * Descriptions with multiple lines are now parsed properly.
  * Tags are now converted into the `list` type.

neosvr_headless_api/response_formats.py
# NeosVR-Headless-API
# Copyright (C) 2022  GlitchFur

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# This file contains format strings that describe how the output of the various
# commands in the headless client should be parsed. If the NeosVR developers
# change the format of any command's response, this file can be easily updated
# so that NeosVR-Headless-API can continue to parse responses properly.

# For more information on how to write these strings, please see:
# https://github.com/r1chardj0n3s/parse
# https://docs.python.org/3/library/string.html#format-string-syntax

# These are for parsing all the startup messages that come BEFORE the prompt.
NEOS_VERSION_FORMAT = "Initializing Neos: {}"
SUPPORTED_TEXTURE_FORMATS_FORMAT = "Supported Texture Formats: {}"
AVAILABLE_LOCALES_FORMAT = "Available locales: {}"
ARGUMENT_FORMAT = "Argument: {}"
COMPATIBILITY_HASH_FORMAT = "Compatibility Hash: {}"
MACHINE_ID_FORMAT = "MachineID: {}"
SUPPORTED_NETWORK_PROTOCOLS_FORMAT = "Supported network protocols: {}"

WORLD_FORMAT = "[{:d}] {name}Users: {users:d}\tPresent: {present:d}\tAccessLevel: {access_level}\tMaxUsers: {max_users:d}"
# Note the lack of space between "ID:" and "{user_id}" in the format string
# below. This is intentional. If the headless user is not logged into a Neos
# account, the ID field will be blank. In this case, "{user_id}" will catch the
# single space character, which we can then interpret as no ID. If we left the
# space in the format string, then `parse()` would fail entirely. This also
# means in all other cases, valid IDs will have an extra space before them, but
# that can be easily trimmed off.
USER_FORMAT = "{name}\tID:{user_id}\tRole: {role}\tPresent: {present}\tPing: {ping:d} ms\tFPS: {fps:g}\tSilenced: {silenced}"
BAN_FORMAT = "[{:d}]\tUsername: {name}\tUserID: {user_id}\tMachineId: {machine_id}"

# STATUS_NAME_FORMAT and STATUS_DESCRIPTION_FORMAT contain "SessionID:" and
# "Tags:" at the end of their format strings respectively because it allows for
# parsing world names and descriptions that contain newline characters. By
# "reading into" the next line of the output like this, we can detect the true
# end of those values.
STATUS_NAME_FORMAT = "Name: {}\nSessionID:{}"
STATUS_SESSION_ID_FORMAT = "SessionID: {}"
STATUS_CURRENT_USERS_FORMAT = "Current Users: {:d}"
STATUS_PRESENT_USERS_FORMAT = "Present Users: {:d}"
STATUS_MAX_USERS_FORMAT = "Max Users: {:d}"
STATUS_UPTIME_FORMAT = "Uptime: {}"
STATUS_ACCESS_LEVEL_FORMAT = "Access Level: {}"
STATUS_HIDDEN_FROM_LISTING_FORMAT = "Hidden from listing: {}"
STATUS_MOBILE_FRIENDLY_FORMAT = "Mobile Friendly: {}"
STATUS_DESCRIPTION_FORMAT = "Description: {}\nTags:{}"
STATUS_TAGS_FORMAT = "Tags: {}"
STATUS_USERS_FORMAT = "Users: {}"


rpc_server.py
#!/usr/bin/env python3

# NeosVR-Headless-API
# Copyright (C) 2022  GlitchFur

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# This is the RPC server which can be used to control headless clients remotely.

import logging
import argparse

from rpyc import Service, restricted
from neosvr_headless_api import HeadlessProcess

logging.basicConfig(
    format="[%(asctime)s][%(levelname)s] %(message)s", level=logging.INFO
)


def main():
    parser = argparse.ArgumentParser(
        description="RPC server for controlling multiple headless clients"
    )
    parser.add_argument(
        "--host",
        default="127.0.0.1",
        help="The host or IP address to bind to. (Default: 127.0.0.1)",
    )
    parser.add_argument(
        "-p",
        "--port",
        type=int,
        default=16881,
        help="The TCP port to bind to. (Default: 16881)",
    )
    args = parser.parse_args()

    from rpyc.utils.server import ThreadedServer

    server = ThreadedServer(
        HeadlessProcessService(), hostname=args.host, port=args.port
    )
    server.start()


class HeadlessProcessService(Service):
    def __init__(self):
        super().__init__()
        self.processes = {}
        self.current_id = 0

    def on_connect(self, conn):
        host, port = conn._config["endpoints"][1]
        logging.info("Client connected from %s:%d" % (host, port))

    def on_disconnect(self, conn):
        host, port = conn._config["endpoints"][1]
        logging.info("Client disconnected from %s:%d" % (host, port))

    def exposed_start_headless_process(self, *args, **kwargs):
        """
        Start a new instance of a `HeadlessProcess`. Returns a tuple in the form
        of (id, `HeadlessProcess` instance). All arguments and keyword arguments
        are passed to `HeadlessProcess` as-is.

        The id can be used to get the `HeadlessProcess` instance again at a
        later time with `get_headless_process()`.
        """
        allowed_access = (
            "neos_dir",
            "config",
            "write",
            "readline",
            "shutdown",
            "sigint",
            "terminate",
            "kill",
            "wait",
        )
        process = restricted(HeadlessProcess(*args, **kwargs), allowed_access)
        self.current_id += 1
        self.processes[self.current_id] = process
        logging.info("Starting headless process with ID: %d" % self.current_id)
        logging.info("Neos Dir: %s" % process.neos_dir)
        logging.info("Config: %s" % process.config)
        logging.info("Total processes running: %d" % len(self.processes))
        return (self.current_id, process)

    def exposed_stop_headless_process(self, pid):
        """
        Stops the `HeadlessProcess` with the given `pid` and removes it from the
        process list. Returns the exit code of the process.
        """
        process = self.processes[pid]
        exit_code = process.shutdown()
        del self.processes[pid]
        logging.info(
            "Headless process with ID %d terminated with return code %d."
            % (pid, exit_code)
        )
        logging.info("Total processes running: %d" % len(self.processes))
        return exit_code

    def exposed_send_signal_headless_process(self, pid, sig):
        """
        Send a signal to the `HeadlessProcess` with the given `pid` and removes
        it from the process list. `sig` is an integer which can be either
        SIGINT (2), SIGTERM (15), or SIGKILL (9). If `sig` is not one of these
        integers, `ValueError` will be raised. Returns the signal used to
        terminate the process as a negative integer.
        """
        process = self.processes[pid]
        if not sig in (2, 9, 15):
            raise ValueError("Signal not allowed: %d" % sig)
        if sig == 2:
            func = process.sigint
        elif sig == 9:
            func = process.kill
        elif sig == 15:
            func = process.terminate
        exit_code = func()
        # TODO: The following code is identical to that of
        # `exposed_stop_headless_process()`
        del self.processes[pid]
        logging.info(
            "Headless process with ID %d terminated with return code %d."
            % (pid, exit_code)
        )
        logging.info("Total processes running: %d" % len(self.processes))
        return exit_code

    def exposed_get_headless_process(self, pid):
        """Returns an existing `HeadlessProcess` with the given `pid`."""
        return self.processes[pid]


if __name__ == "__main__":
    main()
